### PR TITLE
Implement to_df directly in Cython files

### PR DESCRIPTION
**Pull request by timosachsenberg**
Instead of using Python wrapper classes in _dataframes.py to add DataFrame export functionality, implement get_df() directly in the Cython addon files alongside get_data_dict(). This simplifies the architecture by:

- Removing the "duck typing hack" of wrapper classes that override Cython classes
- Putting all DataFrame logic in one place (the Cython addon files)
- Making the code flow more straightforward

Changes:
- MSSpectrum.pyx: Added get_df()
- MSChromatogram.pyx: Added get_df()
- Mobilogram.pyx: Added get_df()
- MSExperiment.pyx: Added get_df_columns(), get_df(), get_ion_df(), get_massql_df()
- ConsensusMap.pyx: Added get_df_columns(), get_intensity_df(), get_metadata_df(), get_df()
- FeatureMap.pyx: Added get_df_columns(), get_df(), get_assigned_peptide_identifications()
- MRMTransitionGroupCP.pyx: New addon with get_chromatogram_df(), get_feature_df() methods
- _dataframes.py: Simplified to only export standalone functions
- README.md: Updated documentation to reflect new pattern

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

## Summary by CodeRabbit

* **New Features**
  * Unified, expanded tabular exports (pandas DataFrame and Apache Arrow) for spectra, chromatograms, features, consensus maps, experiments and identifications, with column discovery, separate intensity/metadata views, selectable columns, and in-place score updates.
* **Documentation**
  * README and docstrings updated with the new DataFrame/column-discovery API and examples; lazy-import guidance for optional dependencies.
* **Bug Fixes**
  * Improved meta-value handling, dtype stability, and clearer errors when optional dependencies are missing.
* **Tests**
  * Extensive tests for DataFrame/Arrow exports, column selection, peptide-ID flows, score updates, and edge cases.
* **Chores**
  * Pandas made optional for install; CI adjusted to install optional deps for testing.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by jpfeuffer** [comment]
Nice! You'll probably need pandas at build time now, not just during testing.
Not sure if we are checking for this in cake or setup.py/project.toml

Bonus points for doing it with arrow, too.

**Comment by jpfeuffer** [comment]
I meant exposing it in arrow Format from c++ for zero copy.
Let's just remove it again. AI is too dumb for this.

Also the interface/parameter names were changed. The tests need to be adapted accordingly.

**Comment by timosachsenberg** [comment]
@copilot    
>       assert exp.get_df(long=True).shape == (10, 4)
               ^^^^^^^^^^^^^^^^^^^^^

tests\unittests\test000.py:3239: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   def get_df(self, columns=None, ms_levels=None, long_format=False):
E   TypeError: get_df() got an unexpected keyword argument 'long'

pyopenms\_pyopenms_4.pyx:7226: TypeError
---------------------------- Captured stdout call -----------------------------

**Comment by timosachsenberg** [comment]
I would say: let's get this merged, as it removes a class of bugs around the duck typing. We can improve / optimize in another PR

File: src/pyOpenMS/addons/ConsensusMap.pyx [hunk 1]
```diff
@@ -163,3 +168,256 @@ from UniqueIdInterface cimport setUniqueId as _setUniqueId
...
+        Useful for discovering available columns before export.
+
+        Args:
+            columns (str): 'default' for standard columns, 'all' for all available columns.
+
+        Returns:
+            list: List of column name strings.
+
+        Example:
+            >>> cmap.get_df_columns()
```
Comment (singjc) [inline src/pyOpenMS/addons/ConsensusMap.pyx:185]: I'm assuming cmap is a module exported from pyopenms? `i.e. import pyopenms.ConsensusMap as cmap`? I would maybe update the Examples to include this, so it's clear where cmap comes from.

Comment (singjc) [reply src/pyOpenMS/addons/ConsensusMap.pyx:185]: Or `cmap` is from a loaded results file?

File: src/pyOpenMS/addons/ConsensusMap.pyx [hunk 2]
```diff
@@ -163,3 +168,256 @@ from UniqueIdInterface cimport setUniqueId as _setUniqueId
...
+
+        Generates a pandas DataFrame with feature intensities from each sample in long format (over files).
+
+        For labelled analyses channel intensities will be in one row, therefore resulting in a semi-long/block format.
+        Resulting DataFrame can be joined with result from get_metadata_df by their index 'id'.
+
+        Returns:
+        pd.DataFrame: intensity DataFrame
+        """
+        import pandas as pd
```
Comment (singjc) [inline src/pyOpenMS/addons/ConsensusMap.pyx]: Can we move the pandas import to the top with numpy? Or is it per defined per function to gate issues with pandas not being installed?

File: src/pyOpenMS/addons/FeatureMap.pyx [hunk 3]
```diff
@@ -80,3 +84,273 @@ from UniqueIdInterface cimport setUniqueId as _setUniqueId
...
+
+        return cols
+
+    def _get_prot_id_filename_from_pep_id(self, pep_id):
+        """
+        _get_prot_id_filename_from_pep_id(self: FeatureMap, pep_id: PeptideIdentification) -> str
+
+        Gets the primary MS run path of the ProteinIdentification linked with the given PeptideIdentification.
+
+        Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/FeatureMap.pyx]:
```suggestion
        Args:
```

File: src/pyOpenMS/addons/FeatureMap.pyx [hunk 4]
```diff
@@ -80,3 +84,273 @@ from UniqueIdInterface cimport setUniqueId as _setUniqueId
...
+
+    def get_df(self, columns=None, meta_values=None, export_peptide_identifications=True):
+        """
+        get_df(self: FeatureMap, columns: Optional[List[str]] = None, meta_values: Optional[Union[List[str], str]] = None, export_peptide_identifications: bool = True) -> pd.DataFrame
+
+        Generates a pandas DataFrame with information contained in the FeatureMap.
+
+        Optionally the feature meta values and information for the assigned PeptideHit can be exported.
+
+        Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/FeatureMap.pyx]:
```suggestion
        Args:
```

File: src/pyOpenMS/addons/FeatureMap.pyx [hunk 5]
```diff
@@ -80,3 +84,273 @@ from UniqueIdInterface cimport setUniqueId as _setUniqueId
...
+                yield from fun(f)
+
+        def extract_meta_data(f):
+            """Extracts feature meta data.
+
+            Extracts information from a given feature with the requested meta values and, if requested,
+            the sequence, score and ID_filename (primary MS run path of the linked ProteinIdentification)
+            of the best PeptideHit (first) assigned to that feature.
+
+            Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/FeatureMap.pyx]:
```suggestion
            Args:
```

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 6]
```diff
@@ -171,3 +173,341 @@
...
+            >>> exp.get_df_columns(long_format=True)
+            ['rt', 'mz', 'intensity', 'ms_level']
+
+            >>> exp.get_df_columns(long_format=False)
+            ['rt', 'ms_level', 'mz_array', 'intensity_array']
+        """
+        if long_format:
+            return ['rt', 'mz', 'intensity', 'ms_level']
+        else:
+            return ['rt', 'ms_level', 'mz_array', 'intensity_array']
```
Comment (singjc) [inline src/pyOpenMS/addons/MSExperiment.pyx:202]: Is it possible to include the scan number / native id? There was an instance when I wanted to extract data for a specific spectrum using the scan number, but the `MSExperiment.get_df` doesn't include this meta data. Doesn't need to be included in this PR, but was just wondering if that would be possible?

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 7]
```diff
@@ -171,3 +173,341 @@
...
+
+    def get_df(self, columns=None, ms_levels=None, long_format=False):
+        """
+        get_df(self: MSExperiment, columns: Optional[List[str]] = None, ms_levels: List[int] = [], long_format: bool = False) -> pd.DataFrame
+
+        Generates a pandas DataFrame with all peaks in the MSExperiment
+
+        Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/MSExperiment.pyx]:
```suggestion
        Args:
```

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 8]
```diff
@@ -171,3 +173,341 @@
...
+        Generates a pandas DataFrame with all peaks and the ion mobility in the MSExperiment
+
+        Returns:
+        pd.DataFrame: feature information stored in a DataFrame
+        """
+        import pandas as pd
+        cols = ["rt", "mz", "intensity", "ion_mobility"]
+        self.updateRanges()
+        spectraarrs2d = self.get2DPeakDataIMLong(self.getMinRT(), self.getMaxRT(), self.getMinMZ(), self.getMaxMZ(), 1)
+        return pd.DataFrame(dict(zip(cols, spectraarrs2d)))
```
[the remark this answers is not in the document]
Comment (singjc) [reply src/pyOpenMS/addons/MSExperiment.pyx:392]: Should the doc-string be updated to reflect this?

File: src/pyOpenMS/addons/MSExperiment.pyx (hunk 6 above)
Comment (singjc) [reply src/pyOpenMS/addons/MSExperiment.pyx:202]: I know the `get_massql_df` has the scan column, but for some reason it seems to require/use more memory.

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 9]
```diff
@@ -171,3 +173,341 @@
...
+        'rt': retention time of the spectrum
+        'polarity': ion mode of the spectrum as integer value (positive: 1, negative: 2)
+        'ion': the ionic mobility of a peak if ion parameter is True
+
+        The MS2 dataframe contains additional columns:
+        'precmz': mass to charge of the precursor ion
+        'ms1scan': number of the corresponding MS1 spectrum
+        'charge': charge of the precursor ion
+
+        Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/MSExperiment.pyx]:
```suggestion
        Args:
```

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 10]
```diff
@@ -171,3 +173,341 @@
...
+        import pandas as pd
+        from . import IonSource as _IonSource
+        self.updateRanges()
+
+        def _get_polarity(spec):
+            '''Returns polarity as an integer value for the massql dataframe.
+
+            According to massql positive polarity is represented by 1 and negative by 2.
+
+            Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/MSExperiment.pyx:441]:
```suggestion
            Args:
```

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 11]
```diff
@@ -171,3 +173,341 @@
...
+
+        def _get_spec_arrays(mslevel):
+            '''Get spectrum data as a matrix.
+
+            Generator yields peak data from each spectrum (with specified MS level) as a numpy.ndarray.
+            Normalized intensity values are calculated and the placeholder values replaced. For 'i_norm' and
+            'i_tic_norm' the intensity values are divided by the maximum intensity value in the spectrum and
+            the sum of intensity values, respectively.
+
+            Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/MSExperiment.pyx:463]:
```suggestion
            Args:
```

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 12]
```diff
@@ -171,3 +173,341 @@
...
+
+        def _get_ion_spec_arrays(mslevel):
+            '''Get spectrum data as a matrix.
+
+            Generator yields peak data from each spectrum (with specified MS level) as a numpy.ndarray.
+            Normalized intensity values are calculated and the placeholder values replaced. For 'i_norm' and
+            'i_tic_norm' the intensity values are divided by the maximum intensity value in the spectrum and
+            the sum of intensity values, respectively.
+
+            Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/MSExperiment.pyx:503]:
```suggestion
            Args:
```

File: src/pyOpenMS/addons/PeptideIdentificationList.pyx [hunk 13]
```diff
@@ -72,3 +75,209 @@
         """
         return f"PeptideIdentificationList(size={self.size()})"
 
+    def get_df(self, decode_ontology=True, default_missing_values=None, export_unidentified=True):
+        """
+        get_df(self: PeptideIdentificationList, decode_ontology: bool = True, default_missing_values: dict = None, export_unidentified: bool = True) -> pd.DataFrame
+
+        Converts the peptide identifications to a pandas DataFrame.
+
+        Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/PeptideIdentificationList.pyx]:
```suggestion
        Args:
```

File: src/pyOpenMS/addons/PeptideIdentificationList.pyx [hunk 14]
```diff
@@ -72,3 +75,209 @@
...
+    def update_scores_from_df(self, df, main_score_name):
+        """
+        update_scores_from_df(self: PeptideIdentificationList, df: pd.DataFrame, main_score_name: str) -> PeptideIdentificationList
+
+        Updates the scores in PeptideIdentification objects using a pandas DataFrame.
+
+        This method is useful when you've computed new scores (e.g., from rescoring)
+        and want to update the original PeptideIdentifications.
+
+        Parameters:
```
Comment (singjc) [inline src/pyOpenMS/addons/PeptideIdentificationList.pyx]:
```suggestion
        Args:
```

**Review by singjc — APPROVED** [review]
Awesome! I like this separation, makes it easier to reference the specific structures dataframe methods.

File: src/pyOpenMS/addons/ConsensusMap.pyx (hunk 1 above)
Comment (timosachsenberg) [reply src/pyOpenMS/addons/ConsensusMap.pyx:185]: It is just a variable name for a consensusmap instance. we can call it in the example how we like. e.g. my_consensus_map

File: src/pyOpenMS/addons/ConsensusMap.pyx (hunk 2 above)
Comment (timosachsenberg) [reply src/pyOpenMS/addons/ConsensusMap.pyx]: yes that was the idea. I am not sure if this is the best way to do it but I think I saw other libraries do it this way. You basically get an import error if it is not installed

File: src/pyOpenMS/addons/MSExperiment.pyx (hunk 6 above)
Comment (timosachsenberg) [reply src/pyOpenMS/addons/MSExperiment.pyx:202]: yes sure. I think we should do that. Right now I did not want to change too much compared to the original implementation. But this is definitely something we can do.
Only caveat: we need to check if this leads to issues in other libraries (e.g. pyopenms-viz) etc. But I think most downstream stuff only looks for the presence of columns and don't mind if a new one is added.

File: src/pyOpenMS/addons/ConsensusMap.pyx (hunk 1 above)
Comment (timosachsenberg) [inline src/pyOpenMS/addons/ConsensusMap.pyx:185]:
```suggestion
            >>> consensusmap.get_df_columns()
```

File: src/pyOpenMS/addons/MSExperiment.pyx (hunk 7 above)
Comment (jpfeuffer) [reply src/pyOpenMS/addons/MSExperiment.pyx]: @timosachsenberg actually I am pretty sure that we are using the default sphinx RST style for docstrings: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format

Comment (jpfeuffer) [reply src/pyOpenMS/addons/MSExperiment.pyx]: @timosachsenberg actually I am pretty sure that we are using the default sphinx RST style for docstrings: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format

**Comment by jpfeuffer** [comment]
- if you want to fail at runtime because pandas is optional, you should remove it from the requirements in setup.py/project.toml etc (one could also think about a "nicer" error)
- the docstrings are all in the wrong format I think. Should be sphinx restructured text. But the first line with the type signature needs to stay (it is used by autodoc)

**Comment by timosachsenberg** [comment]
@copilot to fail at runtime because pandas is optional, you should remove it from the requirements in setup.py/project.toml etc and print a "nicer" error. Also the docstrings 
the docstrings are all in the wrong format. Should be sphinx restructured text. Important: the first line with the type signature needs to stay (it is used by autodoc).

**Comment by timosachsenberg** [comment]
@copilot now the [pyopenms-wheels-and-packages / test (ubuntu-latest) (pull_request)](https://github.com/OpenMS/OpenMS/actions/runs/20390067904/job/58601733946?pr=8520)Failing after 1m action fails because we don't have the optional dependency installed. it is need for running the test

File: src/pyOpenMS/addons/MSExperiment.pyx (hunk 6 above)
Comment (timosachsenberg) [reply src/pyOpenMS/addons/MSExperiment.pyx:202]: In any case: feel free to suggest improvements / extensions. This is still pretty much experimental
